zerver/tornado/event_queue.py

```diff
@@ -1144,7 +1153,10 @@ def get_client_payload(apply_markdown: bool, client_gravatar: bool) -> Dict[str,
             # message data unnecessarily
             continue
 
-        message_dict = get_client_payload(client.apply_markdown, client.client_gravatar)
+        can_access_sender = client.user_profile_id not in user_ids_without_access_to_sender
+        message_dict = get_client_payload(
+            client.apply_markdown, client.client_gravatar, client.realm_string_id, can_access_sender
+        )
```

Comment:
The negative is a good way to do it, to avoid sending extra copies of the whole user ID list.